### DISCUSSION TITLE
Redirect from route handler blocked by CORS policy

### DISCUSSION THREAD

**Discussion by FabioDainese** [vercel/next.js#61451] [Help]
### Summary

So, basically, I've setup a NextJS v14.1.0 to just experiment with various *Stripe* checkout methods, like using *server actions*, *route handlers*, etc. and everything works fine except for the route handler variant, since I get the error `access to fetch at ... from origin ... has been blocked by CORS policy`. 
Now, it makes sense that it raises this issue due to the fact that I'm trying to redirect the user to another separate website - hosted by Stripe - but at the same time then, why the same redirect works on *server actions*? I mean, they are both on the server side and they both redirect to a different website 🧐 What am I missing here?

Route handler:
```typescript
import { redirect } from "next/navigation";

export const POST = async (req: NextRequest) => {
    // Stuff

    return NextResponse.redirect(new URL(checkoutSession.url!), {
        status: 303,
    });

    // Also this `redirect` doesn't work
    // redirect(checkoutSession.url as string);
};
```

Server action:
```typescript
"use server"

import { redirect } from "next/navigation";

export const createCheckoutSession: CreateCheckoutSessionsProps = async (data) => {
    // Stuff

    redirect(checkoutSession.url as string);
};
```

I mean, to fix this, I could simply return the url generated by Stripe - instead of directly performing a redirect - and then redirect the user from the client component with `window.location.replace(...)`, but at the same time I would really like to know why it works on one and not on the other.

Finally, I've also added a sandbox link, so you can easily try this error on a live project (if you fork/download it, keep in mind to also create the `.env.local` file following `.env.example` boilerplate file).

### Example

https://codesandbox.io/p/devbox/infallible-microservice-pkmz9g?file=%2Fapp%2Fapi%2Fstripe%2Fcreate-checkout-session%2Froute.ts%3A42%2C1

**Comment by slava-lu** [comment]
The same problem

**Comment by issam-seghir** [comment]
i have the same issue, any workaround that ?

**Comment by Angryman18** [comment]
facing the very same issue.. any solution yet?

**Comment by Michota** [comment]
same

**Comment by dan-kwiat** [comment]
How are you calling your API route from the frontend?

Instead of this:

```tsx
<button
  onClick={() => {
    fetch("/api/your-endpoint", { method: "POST" })
  }}
>
  Submit
</button>
```

Do this:

```html
<form action="/api/your-endpoint" method="post">
  <button type="submit">Submit</button>
</form>
```

**Comment by ugoi** [comment]
same

**Comment by AndryRoby** [comment]
This isn't the same class of bug as a missing `Access-Control-Allow-Origin` header, it's what happens when `fetch()` follows a cross-origin redirect. When your route handler answers with `NextResponse.redirect(checkoutSession.url)` and your client code calls that route handler with `fetch()` (not a full-page navigation), the browser's default `redirect: 'follow'` behavior means it transparently follows the 307/308 to Stripe's checkout URL, then applies the *same* CORS check to Stripe's response that it applied to your route handler's response. Stripe's checkout page is built to be navigated to directly in the browser, not fetched via JS from an arbitrary origin, so it (correctly, from Stripe's side) sends no `Access-Control-Allow-Origin` permitting your origin to read it via `fetch()`, and the browser reports the whole chain as one CORS failure.

Server actions don't hit this because `redirect()` inside a server action is carried back to the client as a navigation instruction over the action's own response channel; it never asks the browser's `fetch()` to follow an external redirect on its own.

The workaround already mentioned above (return the URL as JSON, then redirect on the client with `window.location.href` or a full-page form submission instead of `fetch()`) is the correct fix, not a hack: once a redirect is meant to leave your app entirely, it needs to be a real browser navigation, not something a cross-origin `fetch()` call follows on your behalf.

A different failure mode from the classic missing-header case, but if it's useful for comparing preflight vs. actual-response headers on any of your other API routes, I built a free client-side tool for that: https://arling.sk/cors-doctor/

**Comment by spideythedev** [comment]
If you're calling the Route Handler with `fetch()` from the browser, a redirect doesn't bypass CORS.

For example, if `/api/foo` redirects to another origin, the browser still applies CORS rules to the redirected request.

If the destination is another origin, make sure that destination returns the appropriate CORS headers, or avoid the cross-origin redirect by handling the request server-side.

If you only need to redirect the browser to another page, use a normal navigation instead of `fetch()`:

```js
window.location.href = "https://example.com";
